tests/test_autoscaler.py

```diff
@@ -0,0 +1,179 @@
+        autoscaler.cooldown_seconds_after_scale_up = 300
+        autoscaler.cooldown_seconds_after_scale_down = 60
+        autoscaler.last_scale_up[app_name] = 1000
+        autoscaler.last_scale_down[app_name] = 800
+        autoscaler._now = Mock(return_value=1200)
+        autoscaler.scale(app)
+        mock_get_statsd_client.return_value.gauge.assert_called_once_with("{}.instance-count".format(app_name), 4)
+        mock_paas_client.return_value.assert_not_called()
+
+    def test_scale_paas_app_fewer_instances_recent_scale_down(self, mock_get_statsd_client, mock_paas_client, *args):
+        app_guid = '11111-11111-11111111-1111'
+        app_name = 'app-name-1'
+        cf_info = {'name': app_name, 'instances': 4, 'guid': app_guid}
+        app = self._get_mock_app(app_name, cf_info)
+        app.get_desired_instance_count = Mock(return_value=3)
+
+        autoscaler = Autoscaler()
+        autoscaler.cooldown_seconds_after_scale_up = 300
+        autoscaler.cooldown_seconds_after_scale_down = 60
+        autoscaler.last_scale_up[app_name] = 1000
+        autoscaler.last_scale_down[app_name] = 1400
+        autoscaler._now = Mock(return_value=1430)
+        autoscaler.scale(app)
+        mock_get_statsd_client.return_value.gauge.assert_called_once_with("{}.instance-count".format(app_name), 4)
+        mock_paas_client.return_value.assert_not_called()
+
+    def test_scale_paas_app_fewer_instances_missing_recent_scale_information(self, mock_get_statsd_client,
+                                                                             mock_paas_client, *args):
+        app_guid = '11111-11111-11111111-1111'
+        app_name = 'app-name-1'
+        cf_info = {'name': app_name, 'instances': 4, 'guid': app_guid}
+        app = self._get_mock_app(app_name, cf_info)
+        app.get_desired_instance_count = Mock(return_value=3)
+
+        autoscaler = Autoscaler()
+        autoscaler.cooldown_seconds_after_scale_up = 300
+        autoscaler.cooldown_seconds_after_scale_down = 60
+        autoscaler._now = Mock(return_value=1000)
+        autoscaler.scale(app)
+        mock_get_statsd_client.return_value.gauge.assert_called_once_with("{}.instance-count".format(app_name), 4)
+        mock_paas_client.return_value.assert_not_called()
+
+
+@patch.object(AwsBaseScaler, '_get_boto3_client')
+@patch('app.autoscaler.PaasClient')
+@patch('app.autoscaler.get_statsd_client')
+@patch.object(ScheduleScaler, '_now', return_value=datetime.datetime(2018, 3, 15, 4, 20, 00))
+@patch.object(ElbScaler, 'gauge')
+@patch.object(ElbScaler, '_get_boto3_client')
+@patch.object(ElbScaler, '_get_request_counts')
+class TestAutoscalerAlmostEndToEnd:
+    def test_scale_up(self,
+                      mock_get_request_counts,
+                      mock_get_boto3_client,
+                      mock_gauge,
+                      mock_schedule_scaler_now,
+                      mock_get_statsd_client,
+                      mock_paas_client,
```

Comment:
pytest-mock gives you the `mocker` fixture, then you can just say

`mock_get_request_counts = mocker.patch.object(ElbScaler, '_get_request_counts')`

etc etc

and avoid having to match up params and decorators